Refactor the project

MosesofEgypt opened the pull request:
* Added launcher logs to ignore 
* Fixed typos in readme, config, and download
* Renamed img\app.ico to img\256x256.ico to simplify code in main
* Simplified and sped up inject_variables in common.py
* Added __all__ to config to clarify module exports
* Fixed missing import of requests library in download
* Made the download increment function a private class method so it
    * can be easily overloaded and doesn't have to be remade each  time download_file is called.
* Added assertion to sha256_hash to make sure a block_size > 0 is given
* Sped up list_files and made it handle receiving directory string  that already have a path separator on the end.

MosesofEgypt committed:
Added launcher logs to ignore
Fixed typos in readme, config, and download
Renamed img\app.ico to img\256x256.ico to simplify code in main
Simplified and sped up inject_variables in common.py
Added __all__ to config to clarify module exports
Fixed missing import of requests library in download
Made the download increment function a private class method so it
    can be easily overloaded and doesn't have to be remade each
    time download_file is called.
Added assertion to sha256_hash to make sure a block_size > 0 is given
Sped up list_files and made it handle receiving directory strings
    that already have a path separator on the end.

james7132 commented on Launcher/config.json:
```diff
@@ -0,0 +1 @@
+{"branches": {"develop": "Development"}, "config_endpoint": "https://patch.houraiteahouse.net/{project}/launcher/config.json", "launcher_endpoint": "https://patch.houraiteahouse.net/{project}/launcher/{platform}/{executable}", "project": "Fantasy Crescendo", "index_endpoint": "https://patch.houraiteahouse.net/{project}/{branch}/{platform}/index.json", "launch_flags": {}, "news_rss_feed": "https:/ ...
```
Add the Launcher folder into .gitignore as well.

james7132 commented on src/config.py:
```diff
@@ -12,18 +12,27 @@
 from util import namedtuple_from_mapping
 from collections import OrderedDict
 
+# there are a lot of module-level declarations below, so to make it
```
__all__ is self-explanatory. Comment is unnecessary.

james7132 commented on src/common.py:
```diff
@@ -24,12 +24,15 @@ def sanitize_url(url):
+
+        if replacement is None:
+            continue
+
```
nit: the extra blank lines here do nothing to help with readability.

james7132 reviewed (approved):
Overall pretty good. Mostly small nitpicks.

james7132 commented:
Looks like the PyInstaller build failed due to the change in icon name.

MosesofEgypt committed:
Changed install/setup.iss to use renamed 256x256.ico
Included CONFIG in src/config.py __all__
Wrote unit tests for util and common

MosesofEgypt committed:
Added launcher and Launcher directories to gitignore
Fixed incorrect icon file name in specs files

MosesofEgypt committed:
Implemented config test, though it cant be run until a bug
    is fixed that is specified in the test
Implemented sha256_hash test for util
Implemented testing inject_variables using a tupperware
Started on test for download module

MosesofEgypt committed:
pep8

MosesofEgypt committed:
Moved attempt to import gettext_windows to where it is used and
    added logic to prevent accessing it if the import failed
Tweaked Download.download_file to allow a comparison hash to be
    provided to the download_file function
Implemented test cases for testing download_file function,
    Download class, and began implementing for test cases for
    the DownloadTracker class

MosesofEgypt committed:
Accidentally broke news feed.
    Will fix later
Started refactoring config module into parameterized functions that
    inject constants to the module at a global level
Refactored main into an initialize function
Tweaked ui.py to access config.py attributes rather than importing
    and using them directly

MosesofEgypt committed:
news still broken.
fixes to prevent accidentally calling setup_directories
    and install_translations more than once

james7132 committed:
Only deploy on master

james7132 committed:
Debug Appveyor deployments

MosesofEgypt committed:
Removed Launcher folder from repo
Added cover folder to gitignore
Added some todo markers
Started on test_ui
Implemented most of test_main
Fixed bugs with _call_download in test_download where it did not actually
    set the data of the session to what was provided
Implemented most tests in test_config and left others as stubs
Removed get_config from config
Added logic around all calls to logger in config to ensure they aren't
    called if the logger isnt setup, otherwise everything will be
    printed out to the console instead of the file

MosesofEgypt committed:
Merge branch 'refactor' of https://github.com/HouraiTeahouse/HouraiLauncher.git into refactor

Commit:
Try turning off view for Linux builds

MosesofEgypt committed:
Refactored icon setting code out of main and into common
Created get_app and get_loop functions for common to allow
    getting the app and/or loop, while creating them if
    they dont already exist
Implemented new tests in test_common for the new functions
Added placeholder call to setup_logger in test_config since
    it no longer spits loads of logging errors if called

MosesofEgypt committed:
Implemented some tests for DownloadTracker
Fixed bugs in test_download tests
Fixed logic bug in config
Fixed bugs in test_common

MosesofEgypt committed:
pep8

MosesofEgypt committed:
Implemented 48% test coverage for ui
Added comments to clarify chmod calls
Refactoring code to be smaller

MosesofEgypt committed:
Attempt to fix a failing unit test on darwin systems

james7132 commented on test/test_ui.py:
```diff
@@ -25,7 +25,8 @@
         launch_flags=dict(Windows=["-bill", "-gates"],
-                          Linux=["-linus", "-torvalds"]),
+                          Linux=["-linus", "-torvalds"],
+                          Darwin=["-linus", "-torvalds"]),
```
Please add a comment here regarding why you added this. Including a link to https://en.wikipedia.org/wiki/Darwin_(operating_system), since it's uncommon to see Darwin as the system platform.

MosesofEgypt committed:
Fixes to prevent crashes and trying to launch binary on
    unsupported systems(Darwin for example)

MosesofEgypt committed:
Another attempt to fix tests failing on other systems by mocking
    os.path.isfile
Added new test to check that the branch does not test to launch
    the binary if the binary file doesn't exist

MosesofEgypt committed:
Implemented test for Branch._diff_files
Refactored Branch._diff_files

MosesofEgypt committed:
Attempt to fix unit test failing on unsupported systems
    Does this by injecting mock values into config for those systems

MosesofEgypt committed:
Implemented get_platform function to alias Darwin systems to OSX
Implemented test for get_platform function
Implemented test_branch_can_preclean_branch_directory
Modified DownloadTrackerMock to support being used in other tests

MosesofEgypt committed:
Attempt to fix a test failure by making sure path separators are all
    backslashes when testing _preclean_branch_directory

MosesofEgypt committed:
Change to os_makedirs_mock to replace all / with \ after using
    os.path.join to concatenate dirs together to ensure \ are used
Added mock for os.path.dirname to make sure dirnames all use \
This would be so much easier if I actually had a mac system to test on

MosesofEgypt committed:
Replaced use of platform.system with util.get_platform in test_common

MosesofEgypt committed:
Implemented async_unittest for writing tests for asynchronous code
Added logging call if main loop is exited unexpectedly

MosesofEgypt committed:
Created async_patch function for easily patching async functions
Fixed some incorrect usage of AsyncMock in AsyncMagicMock

MosesofEgypt committed:
Fixed incorrect usage of run_until_complete in run_async

MosesofEgypt committed:
Implemented asynchronous tests for download.DownloadTracker methods

MosesofEgypt committed:
Implemented async tests for Branch.fetch_remove_index and most
    of the rest of MainWindow's async methods.
Changed SessionMock to be able to be used in a "with" statement
Changed ResponseMock so the json can be explicitely provided

MosesofEgypt committed:
Implemented tests for half of MainWindow.launcher_update_check
Implemented text attribute for ResponseMock

MosesofEgypt committed:
Increased coverage on tests and completed checks on others to
    ensure the tests ran successfully

MosesofEgypt committed:
Fixed missing "self." in ui.launcher_update_check
pep8
Finished unit tests. coverage comes to ~97%(excluding main.py)

james7132 commented on src/common.py:
```diff
@@ -1,35 +1,79 @@
+                config.RESOURCE_DIR, 'img', '%sx%s.ico' % (size, size)),
+            QtCore.QSize(size, size))
+
+    g['app_icon'] = app_icon
```
This doesn't need to be declared in a global context. It's fine leaving it scoped to the function.

james7132 commented on src/config.py:
```diff
@@ -5,97 +5,176 @@
+    if gettext_windows is not None:
+        gettext_windows.setup_env()
+
+    g['TRANSLATIONS'] = gettext.translation(
```
This also could be scoped to the function as well. `TRANSLATIONS.install()` does the desired effect of installing the `_()` builtin for translation.

james7132 commented on src/config.py:
```diff
@@ -5,97 +5,176 @@
+    gettext_windows = None
+    if 'win' in get_platform().lower():
+        if _LOGGER_SETUP:
+            logging.info(
```
If the logger is not setup, this shouldn't throw an error and would effectively be a no-op. Is the _LOGGER_SETUP check necessary?

james7132 commented on src/config.py:
```diff
@@ -5,97 +5,176 @@
+
+    g['_DIRECTORIES_SETUP'] = True
+
+setup_directories()
```
Instead of having these declared in the global context of the module, have main call it instead. This way it's not executed each time it is imported.

james7132 commented on src/ui.py:
```diff
@@ -50,8 +49,12 @@ def index_directory(self):
     def launch_game(self, game_binary, command_args):
         binary_path = os.path.join(self.directory, game_binary)
+        if not os.path.isfile(binary_path):
+            logging.info("Path to binary does not exist: %s" % binary_path)
```
This should be an error log.

james7132 commented on test/async_unittest.py:
```diff
@@ -0,0 +1,57 @@
+from unittest import *  # import everything that IS defined in __all__
+
+
+def AsyncMock(func):
```
Name does not conform to style guide: `async_mock` for functions.

james7132 commented on test/async_unittest.py:
```diff
@@ -0,0 +1,57 @@
+    return mock_coroutine
+
+
+def AsyncMagicMock(*args, **kwargs):
```
Ditto here. MagicMock is a class. AsyncMagicMock is a function: `async_magic_mock`.

james7132 commented on test/test_common.py:
```diff
@@ -0,0 +1,99 @@
+    def test_get_loop_fails_without_app(self):
+        common.app = None
+        common.loop = None
+        try:
```
Use `unittest.TestCase.assertRaises`: https://stackoverflow.com/questions/129507/how-do-you-test-that-a-python-function-throws-an-exception. This provides more accurate data and is more declarative in your test.

james7132 commented on test/test_common.py:
```diff
@@ -0,0 +1,99 @@
+    def test_can_get_app(self):
+        common.app = None
+        app = get_app()
+        self.assertTrue(app)
```
Use unittest.TestCase.assertIsNotNone instead. This explicitly checks for None and cannot be tricked by other falsy results.

james7132 commented on test/test_common.py:
```diff
@@ -0,0 +1,99 @@
+    def test_can_get_loop(self):
+        common.loop = None
+        loop = get_loop()
+        self.assertTrue(loop)
```
Ditto here with the `assertIsNotNone`.

james7132 commented on test/test_common.py:
```diff
@@ -0,0 +1,99 @@
+    def test_cannot_set_icon_without_app(self):
+        common.app = None
+        try:
+            set_app_icon()
```
Ditto here with the `assertRaises`

james7132 commented on test/test_common.py:
```diff
@@ -0,0 +1,99 @@
+        self.assertEqual(len(ICON_SIZES), len(qicon_sizes))
+
+        for q_size in qicon_sizes:
+            self.assertTrue(q_size.height() == q_size.width())
```
Use `assertEquals` instead.

james7132 commented on test/test_config.py:
```diff
@@ -0,0 +1,181 @@
+    raise Exception("The line of code should have not been run.")
+
+
+class RotatingFileHandlerMock(RotatingFileHandler):
```
Instead of creating a dummy class, try patching out the class instead: https://www.relaxdiego.com/2014/04/mocking-objects-in-python.html.

james7132 commented on test/test_config.py:
```diff
@@ -0,0 +1,181 @@
+
+    def test_translations_can_be_installed(self):
+        config.install_translations()
+        self.assertTrue(config._TRANSLATIONS_INSTALLED)
```
Instead of doing this, assert that the `_()` translation function is installed into the global builtins, which it the target goal.

james7132 commented on test/test_config.py:
```diff
@@ -0,0 +1,181 @@
+                read_data=config_test_data)) as m:
+            cfg = config.load_config()
+
+        assert hasattr(config, "TRANSLATION_DIR")
```
DRY: Create a list for these and call it in a loop instead.

MosesofEgypt committed:
Removed the selective import in common for one that might work better
Included error checking when trying to connect online to
    prevent crashes when unable to connect.

james7132 reviewed (changes requested):
Some fixes needed.

MosesofEgypt committed:
Implemented changed suggested here:
https://github.com/HouraiTeahouse/HouraiLauncher/pull/17
